### DISCUSSION TITLE
Log identity provider SAML assertions

### DISCUSSION THREAD

**Discussion by eggyknap** [keycloak/keycloak#13259] [Q&A] [answered]
I'm setting up Keycloak to broker several SAML identity providers, and I would like to set up the mappers to map attributes from these different providers into the linked Keycloak user. But I can't find anywhere to see the SAML data these identity providers send Keycloak, so I can't see what attribute names the providers use. Is there somewhere I can find those assertions, or log them, or something?

**Comment by bonnm** [comment, marked as answer]
I once had the problem to see the assertions Keycloak (in SAML IdP role) is sending to SAML service providers. Setting Keycloak log level to DEBUG helped, to assertions xml was in the logs. Maybe this helps in your scenario. If it does not help and the SAML iDP do not have a documentation, you can play around with the standard attribute map of shibboleth attribute map. Here's the default content, maybe this helps.

`
<Attributes xmlns="urn:mace:shibboleth:2.0:attribute-map" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

    

    

    <Attribute name="urn:mace:dir:attribute-def:eduPersonPrincipalName" id="eppn">
        <AttributeDecoder xsi:type="ScopedAttributeDecoder"/>
    </Attribute>
    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.6" id="eppn">
        <AttributeDecoder xsi:type="ScopedAttributeDecoder"/>
    </Attribute>

    <Attribute name="urn:mace:dir:attribute-def:eduPersonScopedAffiliation" id="affiliation">
        <AttributeDecoder xsi:type="ScopedAttributeDecoder" caseSensitive="false"/>
    </Attribute>
    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.9" id="affiliation">
        <AttributeDecoder xsi:type="ScopedAttributeDecoder" caseSensitive="false"/>
    </Attribute>

    <Attribute name="urn:mace:dir:attribute-def:eduPersonAffiliation" id="unscoped-affiliation">
        <AttributeDecoder xsi:type="StringAttributeDecoder" caseSensitive="false"/>
    </Attribute>
    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.1" id="unscoped-affiliation">
        <AttributeDecoder xsi:type="StringAttributeDecoder" caseSensitive="false"/>
    </Attribute>

    <Attribute name="urn:mace:dir:attribute-def:eduPersonEntitlement" id="entitlement"/>
    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.7" id="entitlement"/>

    

    
    <Attribute name="urn:mace:dir:attribute-def:eduPersonTargetedID" id="targeted-id">
        <AttributeDecoder xsi:type="ScopedAttributeDecoder"/>
        
    </Attribute>

    
    

    
    <Attribute name="urn:oid:1.3.6.1.4.1.5923.1.1.1.10" id="persistent-id">
        <AttributeDecoder xsi:type="NameIDAttributeDecoder" formatter="$NameQualifier!$SPNameQualifier!$Name" defaultQualifiers="true"/>
    </Attribute>

    
    <Attribute name="urn:oasis:names:tc:SAML:2.0:nameid-format:persistent" id="persistent-id">
        <AttributeDecoder xsi:type="NameIDAttributeDecoder" formatter="$NameQualifier!$SPNameQualifier!$Name" defaultQualifiers="true"/>
    </Attribute>

    
    

    
    <Attribute name="urn:mace:dir:attribute-def:cn" id="cn"/>
    <Attribute name="urn:mace:dir:attribute-def:sn" id="sn"/>
    <Attribute name="urn:mace:dir:attribute-def:givenName" id="givenName"/>
    <Attribute name="urn:mace:dir:attribute-def:displayName" id="displayName"/>
    <Attribute name="urn:mace:dir:attribute-def:mail" id="mail"/>
    <Attribute name="urn:mace:dir:attribute-def:telephoneNumber" id="telephoneNumber"/>
    <Attribute name="urn:mace:dir:attribute-def:title" id="title"/>
    <Attribute name="urn:mace:dir:attribute-def:initials" id="initials"/>
    <Attribute name="urn:mace:dir:attribute-def:description" id="description"/>
    <Attribute name="urn:mace:dir:attribute-def:carLicense" id="carLicense"/>
    <Attribute name="urn:mace:dir:attribute-def:departmentNumber" id="departmentNumber"/>
    <Attribute name="urn:mace:dir:attribute-def:employeeNumber" id="employeeNumber"/>
    <Attribute name="urn:mace:dir:attribute-def:employeeType" id="employeeType"/>
    <Attribute name="urn:mace:dir:attribute-def:preferredLanguage" id="preferredLanguage"/>
    <Attribute name="urn:mace:dir:attribute-def:manager" id="manager"/>
    <Attribute name="urn:mace:dir:attribute-def:seeAlso" id="seeAlso"/>
    <Attribute name="urn:mace:dir:attribute-def:facsimileTelephoneNumber" id="facsimileTelephoneNumber"/>
    <Attribute name="urn:mace:dir:attribute-def:street" id="street"/>
    <Attribute name="urn:mace:dir:attribute-def:postOfficeBox" id="postOfficeBox"/>
    <Attribute name="urn:mace:dir:attribute-def:postalCode" id="postalCode"/>
    <Attribute name="urn:mace:dir:attribute-def:st" id="st"/>
    <Attribute name="urn:mace:dir:attribute-def:l" id="l"/>
    <Attribute name="urn:mace:dir:attribute-def:o" id="o"/>
    <Attribute name="urn:mace:dir:attribute-def:ou" id="ou"/>
    <Attribute name="urn:mace:dir:attribute-def:businessCategory" id="businessCategory"/>
    <Attribute name="urn:mace:dir:attribute-def:physicalDeliveryOfficeName" id="physicalDeliveryOfficeName"/>

    <Attribute name="urn:oid:2.5.4.3" id="cn"/>
    <Attribute name="urn:oid:2.5.4.4" id="sn"/>
    <Attribute name="urn:oid:2.5.4.42" id="givenName"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.241" id="displayName"/>
    <Attribute name="urn:oid:0.9.2342.19200300.100.1.3" id="mail"/>
    <Attribute name="urn:oid:2.5.4.20" id="telephoneNumber"/>
    <Attribute name="urn:oid:2.5.4.12" id="title"/>
    <Attribute name="urn:oid:2.5.4.43" id="initials"/>
    <Attribute name="urn:oid:2.5.4.13" id="description"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.1" id="carLicense"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.2" id="departmentNumber"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.3" id="employeeNumber"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.4" id="employeeType"/>
    <Attribute name="urn:oid:2.16.840.1.113730.3.1.39" id="preferredLanguage"/>
    <Attribute name="urn:oid:0.9.2342.19200300.100.1.10" id="manager"/>
    <Attribute name="urn:oid:2.5.4.34" id="seeAlso"/>
    <Attribute name="urn:oid:2.5.4.23" id="facsimileTelephoneNumber"/>
    <Attribute name="urn:oid:2.5.4.9" id="street"/>
    <Attribute name="urn:oid:2.5.4.18" id="postOfficeBox"/>
    <Attribute name="urn:oid:2.5.4.17" id="postalCode"/>
    <Attribute name="urn:oid:2.5.4.8" id="st"/>
    <Attribute name="urn:oid:2.5.4.7" id="l"/>
    <Attribute name="urn:oid:2.5.4.10" id="o"/>
    <Attribute name="urn:oid:2.5.4.11" id="ou"/>
    <Attribute name="urn:oid:2.5.4.15" id="businessCategory"/>
    <Attribute name="urn:oid:2.5.4.19" id="physicalDeliveryOfficeName"/>
`

**Reply by eggyknap** [reply]
That helped quite a lot. Specifically, setting the `org.keycloak.saml.SAMLRequestParser` logging category to `DEBUG` level gets me the SAML assertion in the logs. Thanks!

**Comment by christian-2** [comment]
@bonnm I do have a Keycloak 21.1.1 server with a functioning SAML client, i.e. my Keycloak works as SAML IdP alright. I also have `log-level=DEBUG`. Yet I do not see any occurrences of `Assertion` (as case-insensitive string; other than the attribute `AssertionConsumerServiceURL` in relation to `AuthnRequest`) in Keycloak's log. That seems to run counter to your previous answer. 

What could I be missing? BTW I also need to inspect those assertions for debugging attribute mapping.

**Reply by christian-2** [reply]
So I have been able to glean the answer from the Keycloak source code: [`SAMLRequestParser`](https://github.com/keycloak/keycloak/blob/main/saml-core/src/main/java/org/keycloak/saml/SAMLRequestParser.java#L59-L60) does log `AuthRequest` at log level `DEBUG` ("SAML Redirect Binding ..."); [`SAML2LoginResponseBuilder`](https://github.com/keycloak/keycloak/blob/main/saml-core/src/main/java/org/keycloak/saml/SAML2LoginResponseBuilder.java#L182) does log `Response` (with `Assertion`) at log level `TRACE` ("SAML Response Document ..."). Hence `log-level=DEBUG` is not sufficient for logging both; either [`TRACE` or `ALL`](https://www.keycloak.org/server/logging) is required.

**Reply by madsboddum** [reply]
> So I have been able to glean the answer from the Keycloak source code: [`SAMLRequestParser`](https://github.com/keycloak/keycloak/blob/main/saml-core/src/main/java/org/keycloak/saml/SAMLRequestParser.java#L59-L60) does log `AuthRequest` at log level `DEBUG` ("SAML Redirect Binding ..."); [`SAML2LoginResponseBuilder`](https://github.com/keycloak/keycloak/blob/main/saml-core/src/main/java/org/keycloak/saml/SAML2LoginResponseBuilder.java#L182) does log `Response` (with `Assertion`) at log level `TRACE` ("SAML Response Document ..."). Hence `log-level=DEBUG` is not sufficient for logging both; either [`TRACE` or `ALL`](https://www.keycloak.org/server/logging) is required.

This helped me enable printing of assertions - thanks!

**Comment by dteleguin** [comment]
If everything happens in the front channel, you can simply grab SAML payloads from your browser's Dev Tools -> Network, or use a browser extension like [SAML Tracer](https://addons.mozilla.org/ru/firefox/addon/saml-tracer/). You can also use [online SAML tools](https://www.samltool.com/online_tools.php) to decode SAML messages.

**Reply by marcosadsj** [reply]
Utilizing browser extensions can be quite dangerous. I've used the https://samltool.io/ provided by Auth0, but my genuine recommendation is to filter browser requests for 'saml' and locate the 'SamlResponse' field in the payload request. Copy and decode it using a base64 tool, and then format the XML result for better comprehension.
